crates/bevy_asset/src/processor/tests.rs

```diff
@@ -113,14 +179,25 @@ fn create_app_with_asset_processor() -> AppWithProcessor {
         .set_log_factory(Box::new(FakeTransactionLogFactory))
         .unwrap();
 
+    // Now that we've built the app, finish all the processing dirs.
+
     AppWithProcessor {
         app,
-        source_dir,
-        processed_dir,
+        default_source_dirs: default_source_dirs.finish(),
+        extra_sources_dirs: extra_sources_dirs
+            .into_iter()
+            .map(|(name, dirs)| (name, dirs.finish()))
+            .collect(),
     }
 }
 
 fn run_app_until_finished_processing(app: &mut App) {
+    // If the original source changes through an AssetSourceEvent, we'll be racing (on
+    // multithreaded) between this and processor thread switching the state to `Processing`. So do a
+    // fixed number of iterations so the processor thread is likely to win.
+    for _ in 0..5 {
+        app.update();
+    }
```

Comment:
If this is at all flaky it's going to be really annoying for everyone. If it is flaky then we should find some other way of preventing this race, if it's not flaky I think the comment should be reworded to explain why.